Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,370 @@
+A macro is a convenient way to call a rule, and not recognized by Bazel in the
+same way a rule is.
+
+_Note: The main distinction between a `.bzl` and a `BUILD` file is `BUILD` files are
+used to create targets by calling macros and rules. `.bzl` files define the
+implementation._
+
+To create a wrapper for `objc_library`, create the file `objc_library.bzl`. The
+following macro restricts the customization, and enforces defaults of the
+native `objc_library` rule.
+
+```
+def objc_library(name, srcs=[], hdrs=[], deps=[], data=[]):
+    """
+    An objc_library that turns on pedantic warnings,
+    and enables modules
+    """
+    native.objc_library(
+        name=name,
+        deps=deps,
+        data=data,
+        srcs=srcs,
+        hdrs=hdrs,
+        enable_modules=True,
+        copts=[
+            "-Wpedantic"
+        ]
+    )
+```
+
+By loading our `objc_library` into a `BUILD` file, it will override the native
+`objc_library` rule, which is automatically imported.
+```
+load(":objc_library.bzl", "objc_library")
+
+# This calls the custom objc_library macro. Thankfully, building this with
+# Bazel will error out due to the unavailable argument `copts` used here.
+objc_library(name="some", copts=["-DSOME"])
+```
+
+The same principals can be applied to many rules: wrapping macros with macros.
+With Starlark, the possibilities are endless! _technically they are endless as
+the language is not turing complete_. Please see the
+[`objc_library`](https://docs.bazel.build/versions/master/be/objective-c.html)
+documentation for all possible arguments. _Note: Unlike Starlark rules which
+are easily added on to Bazel, the `objc_library` is part of the internal java
+rules shipped with the Bazel
+binary._ 
+
+### Toolchains
+
+In additional to configuring the build with rules, bazel provides a additional
+primitive, the toolchain. Toolchains provide default compilers and arguments
+for those compilers. For the native c++ rules, this is relied upon to configure
+the many flags required for cross compilation.
+
+For most iOS projects, configuring the c++ toolchain isn't required. Defining a
+custom toolchain is useful to gives full control over compiler invocations or
+to customize the compiler being invoked.  Please see the [toolchain
+documentation](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html)
+to learn more.
+
+### Generated Xcode projects
+
+In an Xcode world, folks checked in a project which contains an listing of
+files, build settings, and IDE state. When the project scales, the project file
+model breaks down. Auditing and code reviewing config changes in Xcode projects
+quickly becomes difficult. With Bazel, human readable `BUILD` files are the
+source of truth for build and Xcode configuration. Tools like
+[XCHammer](https://github.com/pinterest/xchammer), and
+[Tulsi](https://github.com/bazelbuild/tulsi) use an aspect to traverse the
```

Comment:
Whats an Aspect? Should we cover that here or at least link to it?

Reply:
Liked up a pretty good post about aspects: https://kchodorow.com/2017/01/10/aspects-the-fan-fic-of-build-rules/